test-loop-tests/src/examples/validator_rotation.rs

```diff
@@ -0,0 +1,183 @@
+            .map(|v| {
+                let (account, _, _) = v.destructure();
+                account
+            })
+            .sorted()
+            .collect();
+        let validators: Vec<_> = validators.iter().map(|v| v.account.clone()).sorted().collect();
+        assert_eq!(current_validators, validators);
+    };
+
+    let stake_original = |env: &mut TestLoopEnv| {
+        let mut stake_txs: Vec<_> =
+            original_validators.iter().map(|v| v.stake_tx(env, stake)).collect();
+        stake_txs.extend(alternative_validators.iter().map(|v| v.stake_tx(env, 0)));
+        transactions::run_txs_parallel(
+            &mut env.test_loop,
+            stake_txs,
+            &env.node_datas,
+            Duration::seconds(5),
+        );
+    };
+
+    let stake_alternative = |env: &mut TestLoopEnv| {
+        let mut stake_txs: Vec<_> =
+            alternative_validators.iter().map(|v| v.stake_tx(env, stake)).collect();
+        stake_txs.extend(original_validators.iter().map(|v| v.stake_tx(env, 0)));
+        transactions::run_txs_parallel(
+            &mut env.test_loop,
+            stake_txs,
+            &env.node_datas,
+            Duration::seconds(5),
+        );
+    };
+
+    let run_until_next_epoch = |env: &mut TestLoopEnv| {
+        let next_epoch_id =
+            env.test_loop.data.get(&client_actor_handle).client.chain.head().unwrap().next_epoch_id;
+
+        env.test_loop.run_until(
+            |test_loop_data| {
+                let client = &test_loop_data.get(&client_actor_handle).client;
+                let epoch_id = client.chain.head().unwrap().epoch_id;
+                epoch_id == next_epoch_id
+            },
+            Duration::seconds(1 + i64::try_from(epoch_length).unwrap()),
+        );
+    };
+
+    assert_current_validators(&env, &original_validators);
+
+    // At the end of epoch T we define epoch_info (which defines validator set) for epoch T+2 using
+    // current staking. Because of this validator reassignment would have 1 epoch lag compared to
+    // staking distribution.
+
+    stake_alternative(&mut env);
+    run_until_next_epoch(&mut env);
+    assert_current_validators(&env, &original_validators);
+
+    stake_original(&mut env);
+    run_until_next_epoch(&mut env);
+    assert_current_validators(&env, &alternative_validators);
+
+    stake_alternative(&mut env);
+    run_until_next_epoch(&mut env);
+    assert_current_validators(&env, &original_validators);
+
+    stake_original(&mut env);
+    run_until_next_epoch(&mut env);
+    assert_current_validators(&env, &alternative_validators);
+
+    env.shutdown_and_drain_remaining_events(Duration::seconds(10));
+}
+
+struct Validator {
+    account: AccountId,
+    signer: Signer,
+}
+
+impl Validator {
+    fn stake_tx(&self, env: &mut TestLoopEnv, stake: Balance) -> SignedTransaction {
+        let block_hash = transactions::get_shared_block_hash(&env.node_datas, &env.test_loop.data);
+        SignedTransaction::from_actions(
```

Comment:
nit: can we use `SignedTransaction::stake` here?

Reply:
Good idea, done in f40ef4b